pkg/ebpf/verifier/stats.go

```diff
@@ -154,39 +145,47 @@ func generateLoadFunction(file string, stats map[string]*Statistics, failedToLoa
 			strings.Split(filepath.Base(file), ".")[0], "-", "_",
 		)
 		for _, progSpec := range collectionSpec.Programs {
-			prog := interfaceMap[fmt.Sprintf("%s_%s", progSpec.Name, objectFileName)]
-			err = collectionSpec.LoadAndAssign(prog, &opts)
+			prog := reflect.New(
+				reflect.StructOf([]reflect.StructField{
+					{
+						Name: fmt.Sprintf("Func_%s", progSpec.Name),
+						Type: reflect.TypeOf(&ebpf.Program{}),
+						Tag:  reflect.StructTag(fmt.Sprintf(`ebpf:"%s"`, progSpec.Name)),
+					},
+				}),
+			)
+			err = collectionSpec.LoadAndAssign(prog.Elem().Addr().Interface(), &opts)
 			if err != nil {
 				log.Printf("failed to load and assign ebpf.Program in file %s: %v", objectFileName, err)
-				failedToLoad[programKey(progSpec.Name, objectFileName)] = struct{}{}
+				failedToLoad[fmt.Sprintf("%s/%s", objectFileName, progSpec.Name)] = struct{}{}
 				continue
 			}
 
-			progPtr := reflect.ValueOf(prog)
-			if progPtr.Type().Kind() != reflect.Ptr {
+			if prog.Type().Kind() != reflect.Ptr {
 				return fmt.Errorf("%T is not a pointer to struct", prog)
 			}
 
-			if progPtr.IsNil() {
-				return fmt.Errorf("nil pointer to %T", progPtr)
+			if prog.IsNil() {
+				return fmt.Errorf("nil pointer to %T", prog)
 			}
 
-			progElem := progPtr.Elem()
+			progElem := prog.Elem()
 			if progElem.Kind() != reflect.Struct {
 				return fmt.Errorf("%s is not a struct", progElem)
 			}
 			for i := 0; i < progElem.NumField(); i++ {
-				programName := progElem.Type().Field(i).Name
-
 				field := progElem.Field(i)
 				switch field.Type() {
 				case reflect.TypeOf((*ebpf.Program)(nil)):
 					p := field.Interface().(*ebpf.Program)
+					// All unassigned programs and maps are cleaned up by the ebpf loader: https://github.com/cilium/ebpf/blob/main/collection.go#L439
+					// We only need to take care to cleanup assigned programs.
+					defer p.Close()
```

Comment:
This will `defer` to the end of the function, not the loop iteration. Do you want it sooner?